lang-ref/part1.md

```diff
@@ -0,0 +1,941 @@
+
+```
+type StructuredName record {
+    string firstName;
+    string lastName;
+};
+
+type Name StructuredName|string;
+```
+ 
+At runtime, you can define the business logic by checking the currently held value of the union using the `is` operator.
+
+```
+function nameToString(Name nm) returns string {
+   
+    if nm is string {
+        return nm;
+        } else {
+        return nm.firstName + " " + nm.lastName;
+    }
+}
+```
+ 
+In this way, the ``is`` operator in a condition causes the declared type to be narrowed.<br><br>
+
+## Error Reporting
+
+Ballerina does not support the notion of exception. Instead, errors are handled as part of the normal control flow. Errors are first-class citizens in Ballerina. There are a few mechanisms for handling errors, which are centered around an error type. error is a basic type in Ballerina, and error values can be passed around just like any other type.
+
+Errors are reported by functions returning error values.
+
+```
+function parse(string s) returns int|error {
+    
+    int n = 0;
+    int[] cps = s.toCodePointInts();
+    
+    foreach int cp in cps {
+        int p = cp - 0x30;
+        if p < 0 || p > 9 {
+            return error("not a digit");
+        }
+        n = n * 10 + p;
+    }
+    return n;
+}
+```
+ 
+In the above code example, the function **``parse()``** returns a union of integer or error type. At the point of returning the error, the function *``error(“not a digit”)``*  is called.
+
+In case a function is expected to return only error values, a return type of ``error?`` is used.   Ignoring an error also has to be done explicitly.
+
+An error value includes a string message and the stack trace from the point where *``error( )``* is called, and it is immutable.<br><br>
+
+## Error Handling
+
+When an error is reported from a function, it is passed up to its caller. Ultimately it is handled at the main function, which can return it. Alternatively, you can use the ``is`` operator to check for the error type as part of the regular control flow to make a decision.
+
+```
+function intFromBytes(byte[] bytes) returns int|error {
+    
+    string|error ret = string:fromBytes(bytes);
+
+    if ret is error {
+        return ret;
+    } else {
+        return int:fromString(ret);
+    }
+}
+```
+
+So in the above code example, the function defines a local variable **``ret``** to hold the returned value of function **``fromBytes( )``**. **``ret``** is a union of string or error type. Therefore to ascertain the actual type, ``is`` operator is used.<br><br>
+
+## ``check`` expression
+
+Instead of using the ``is`` operator within the regular conditional flow for checking errors, Ballerina has a shorthand method available.  Using the ``check`` keyword, error handling and return statements are much more concise.
```

Comment:
```suggestion
Instead of using the ``is`` operator within the regular conditional flow for checking errors, Ballerina has a shorthand method available. Using the ``check`` keyword, error handling and return statements are much more concise.
```